Define new tag for atomic ABI (#385)
riscv-non-isa/riscv-elf-psabi-doc: riscv-elf.adoc

[[EF_RISCV_RVE]]
  EF_RISCV_RVE (0x0008)::: This bit is set when the binary targets the E ABI.

  EF_RISCV_ATOMIC_ABI_UNKNOWN (0x0000):::
  EF_RISCV_ATOMIC_ABI_TSO (0x0010):::
  EF_RISCV_ATOMIC_ABI_A6C (0x0020):::
  EF_RISCV_ATOMIC_ABI_A6S (0x0030):::
  These flags identify the atomic ABI in use for this ELF file. For detailed
  mapping, refer to the _RISC-V atomics mappings_, 0x0040, 0x0050, 0x0060 and
  0x0070 are reserved for future atomic ABIs.

  EF_RISCV_ATOMIC_ABI (0x0070)::: This macro is used as a mask to test for one

>   of the above floating-point ABIs, e.g.,

Reviewer: s/floating-point/atomic/

| Tag_RISCV_priv_spec                 |        8 | uleb128        | *Deprecated*, indicates the major version of the privileged specification.
| Tag_RISCV_priv_spec_minor           |       10 | uleb128        | *Deprecated*, indicates the minor version of the privileged specification.
| Tag_RISCV_priv_spec_revision        |       12 | uleb128        | *Deprecated*, indicates the revision version of the privileged specification.

> | Tag_RISCV_atomic_abi                |       14 | uleb128        | Indicates for the atomic ABI.

Reviewer 2: ```suggestion
| Tag_RISCV_atomic_abi                |       14 | uleb128        | Indicates the atomics ABI.
```

or if it's more normative to be verbose, then maybe:

```suggestion
| Tag_RISCV_atomic_abi                |       14 | uleb128        | Indicates which version of the atomics ABI is being used.
```

The linker should report errors if object files of different privileged
specification versions are merged.

===== Tag_RISCV_atomic_abi, 14, uleb128=version

Tag_RISCV_atomic_abi denotes the atomic ABI used within this object
file. Its values are defined as follows:

.RISC-V attributes
[cols="4,>2,2,5"]
[width=100%]
|===
| Value | Symbolic Name       | Compatibility                        | Description

| 0     | UNKNOWN_ATOMIC_ABI  | Compatible with all other value.     | This object use unknown atomic ABI.
| 1     | A6C_ATOMIC_ABI      | Compatible with unknown and A6S_ATOMIC_ABI. | This object use the A6 classical atomic ABI, which defined in table A.6 in <<riscv-unpriv-20191213>>.

> | 2     | A6S_ATOMIC_ABI      | Compatible with unknown and A6C_ATOMIC_ABI. | This object use the strengthened A6 ABI, which defined in <<Mappings from C/C++ primitives to RISC-V primitives>>.

Reviewer 3: I still think we need a way to distinguish <uses the new strengthened mappings, but is also compatible with A.6 classic; does not use Note 3 mappings> from <uses the new strengthened mappings, but is no longer compatible with A.6 classic, because it uses one of the two Note 3 mappings>. In my view, values 1 and 2 should be orthogonal; you can be compatible with one or both, and all three options will matter in the short term. One of those Note 3 mappings is possible now, and is probably beneficial on systems that never supported A.6 classic, and don't care much about imported A.6 classic object files.

Author: Does define A7_ATOMIC_ABI address with the merge rule describe in https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/391#issuecomment-1644393301 is address your concern here?

Reviewer 3: Yes. We need A6C_ATOMIC_ABI, A7_ATOMIC_ABI, and "compatible with both", which is A6S_ATOMIC_ABI. I'm fine with Patrick's proposal in #391 .

Author: @hboehm @patrick-rivos any suggestion on how to describe A7_ATOMIC_ABI?, `This object use the A7 ABI, which defined in <<Mappings from C/C++ primitives to RISC-V primitives>> with note 3.`?

Reviewer 4: I took a shot at it - it's a bit clunky so feel free to change it.
```
This object uses the A7 ABI, which is defined as a set of mappings from <<Mappings from C/C++ primitives to RISC-V primitives>> with one or more note 3 annotated mappings.
```

Reviewer 3: That seems close to me. Especially since A7 is already in the name, can we just say:

Uses <<Mappings from C/C++ primitives to RISC-V primitives>> and may rely on one or more note 3 annotated mapping. ?

I expect in the future a bunch of implementations will just use this without checking whether or not such a mapping was actually used.

The A6S mapping would be the same, but say "does not rely on note 3 annotated mappings".

Alternatively, we may want to define some alternate term to replace "note 3 mappings", like "A6C-incompatible mappings".

Author: Added A7_ATOMIC_ABI, and let me know if it OK to you and then I gonna prepare binutils PoC, it's kind of process of psABI, then we can land this :)

Reviewer 4: This looks good to me!

Reviewer 3: I made some trivial suggestions elsewhere. But the substance looks good to me as well.

Author: @hboehm thanks, suggestions applied :)

The linker should report errors if object files of different privileged
specification versions are merged.

===== Tag_RISCV_atomic_abi, 14, uleb128=version

> Tag_RISCV_atomic_abi denotes the atomic for this object

Reviewer 2: ```suggestion
Tag_RISCV_atomic_abi denotes the atomic ABI used within this object
```

file. Its values are defined as follows:

.RISC-V attributes
[cols="4,>2,2,5"]
[width=100%]
|===
| Value | Symbolic Name       | Compatibility                        | Description

> | 0     | UNKNOWN_ATOMIC_ABI  | Compatible with all other value.     | This object use unknown atomic ABI.

Reviewer 2: Should this be compatible w/ all other values? Anything that predates support for this tag will fall into this category, correct? So that would be the current LLVM behavior(A6C), the old GCC behavior(some kind of broken), and the new GCC behavior(A6S), right? 

These compatibility concerns have been brought up a few times, so maybe it would be good to outline the choices we've arrived at in a note with our rationale?

Author: Your understanding is right, That's good point, we should have some note to mention that!

The linker should report errors if object files of different privileged
specification versions are merged.

===== Tag_RISCV_atomic_abi, 14, uleb128=version

Tag_RISCV_atomic_abi denotes the atomic ABI used within this object
file. Its values are defined as follows:

[cols="1,2,5"]
[width=100%]
|===
| Value | Symbolic Name       | Description

> | 0     | UNKNOWN_ATOMIC_ABI  | This object use unknown atomic ABI.

Reviewer 5: Normally constants would have a common prefix not a common suffix. But in this case why are we even including the prefix/suffix?

Author: Good point, it already under `Tag_RISCV_atomic_abi`, so prefix/suffix seems redundant.

| 3     | A7_ATOMIC_ABI       | This object use the A7 atomic ABI, which uses the atomic mapping defined by <<Mappings from C/C++ primitives to RISC-V primitives>> and may rely on note 3 annotated mapping.
|===

Merge policy:::
The linker should report errors if object files of incompatible atomic ABI
beginning merged, the compatibility rules for atomic ABIs can be found in
compatibility column in the following table.

[cols="6,2,3"]
[width=100%]
|===
| Input Values                              | Compatible? | Ouput Value

| UNKNOWN_ATOMIC_ABI and A6C_ATOMIC_ABI     | Yes         | A6C_ATOMIC_ABI
| UNKNOWN_ATOMIC_ABI and A6S_ATOMIC_ABI     | Yes         | A6S_ATOMIC_ABI
| UNKNOWN_ATOMIC_ABI and A7_ATOMIC_ABI      | Yes         | A7_ATOMIC_ABI
| A6C_ATOMIC_ABI and A6S_ATOMIC_ABI         | Yes         | A6C_ATOMIC_ABI
| A6C_ATOMIC_ABI and A7_ATOMIC_ABI          | No          | -
| A6S_ATOMIC_ABI and A7_ATOMIC_ABI          | Yes         | A7_ATOMIC_ABI
|===

NOTE: Merging object files with the same ABI will result in the same ABI.

NOTE: Program built without A-extension is classified into UNKNOWN_ATOMIC_ABI

> now, and we intend to define new value for those case in future.

Reviewer 3: Maybe rephrase this as:
"Programs that implement atomic operations without relying in the A-extension are classified as
UNKNOWN_ATOMIC_ABI for now. A new value for those may be defined in the future."

Author: Applied, thanks for suggestion :)

The linker should report errors if object files of different privileged
specification versions are merged.

===== Tag_RISCV_atomic_abi, 14, uleb128=version

Tag_RISCV_atomic_abi denotes the atomic ABI used within this object
file. Its values are defined as follows:

[cols="1,2,5"]
[width=100%]
|===
| Value | Symbolic Name | Description

| 0     | UNKNOWN  | This object uses unknown atomic ABI.
| 1     | A6C      | This object uses the A6 classical atomic ABI, which is defined in table A.6 in <<riscv-unpriv-20191213>>.
| 2     | A6S      | This object uses the strengthened A6 ABI, which uses the atomic mapping defined by <<Mappings from C/C++ primitives to RISC-V primitives>> and does not rely on any note 3 annotated mappings.
| 3     | A7       | This object uses the A7 atomic ABI, which uses the atomic mapping defined by <<Mappings from C/C++ primitives to RISC-V primitives>> and may rely on note 3 annotated mappings.
|===

Merge policy:::

> The linker should report errors if object files of incompatible atomic ABI

Reviewer 5: ```suggestion
The linker should report errors if object files with incompatible atomics ABIs
```

are merged; the compatibility rules for atomic ABIs can be found in the
compatibility column in the following table.

[cols="6,2,3"]
[width=100%]
|===
| Input Values        | Compatible? | Ouput Value

| UNKNOWN and A6C     | Yes         | A6C
| UNKNOWN and A6S     | Yes         | A6S
| UNKNOWN and A7      | Yes         | A7
| A6C and A6S         | Yes         | A6C
| A6C and A7          | No          | -
| A6S and A7          | Yes         | A7
|===

NOTE: Merging object files with the same ABI will result in the same ABI.

> NOTE: Programs that implement atomic operations without relying in the

Reviewer 5: ```suggestion
NOTE: Programs that implement atomic operations without relying on the
```
